### PR TITLE
sock: doc fixes

### PR DESCRIPTION
**Pull request by miri64**
Fixes the comments in #5758.

### CODE REVIEW

**Comment by miri64** [comment]
Also includes the HOWTOs requested by @emmanuelsearch.

File: sys/include/net/sock.h [hunk 1]
```diff
@@ -34,39 +33,55 @@
  *    +---------------+
  * ~~~~~~~~~~~~~~~~~~~~
  *
- * This module provides a minimal set of functions to establish connections or
- * send and receives datagrams using different types of communication. Together,
- * they serve as a common API that connects application- and network stack code.
+ * This module provides a set of functions to establish connections or send and
+ * receive datagrams using different types of protocols. Together, they serve as
+ * an API allows an application or library to connect to a network.
```
Comment (aabadie) [inline sys/include/net/sock.h]: `that allows` or `allowing`

File: sys/include/net/sock.h [hunk 2]
```diff
@@ -34,39 +33,55 @@
...
  * the network stack underneath to be switched simply by changing the
  * `USEMODULE` definitions in the application's Makefile.
  *
+ * The actual code very much depends on the used `sock` type. Please refer to
+ * their documentation for specific examples.
+ *
+ * Implementors Notes
+ * ==================
+ * ### Type definition
+ * For simplicity and modularity this API doesn't put any restriction of the
```
Comment (aabadie) [inline sys/include/net/sock.h]: on the

File: sys/include/net/sock.h [hunk 3]
```diff
@@ -34,39 +33,55 @@
...
+ * actual implementation of the type. For example, one implementation might
+ * choose to have all `sock` types have a common base class or use the raw IP
```
Comment (aabadie) [inline sys/include/net/sock.h]: types having

File: sys/include/net/sock/ip.h [hunk 4]
```diff
@@ -12,7 +12,244 @@
...
+ * ### A Simple IPv6 Client
+ * There are two kinds of clients. Those that do expect a reply and those who
+ * don't. A client that does not require a reply is very simple to implement in
+ * one line:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * res = sock_ip_send(NULL, data, data_len, PROTNUM, &remote);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * With `data` being the data send, `data_len` the length of `data`, `PROTNUM`
```
Comment (aabadie) [inline sys/include/net/sock/ip.h]: data sent

File: sys/include/net/sock/ip.h [hunk 5]
```diff
@@ -12,7 +12,244 @@
...
+ * the next header number for the send packet and `remote` the remote end point
```
Comment (aabadie) [inline sys/include/net/sock/ip.h]: sent packet

File: sys/include/net/sock/ip.h [hunk 6]
```diff
@@ -12,7 +12,244 @@
...
+ * the packet is to be sent.
```
Comment (aabadie) [inline sys/include/net/sock/ip.h]: that is to be

File: sys/include/net/sock/ip.h [hunk 7]
```diff
@@ -12,7 +12,244 @@
...
+ *
+ * To see some other utilization we look at a more complex example in form of
```
Comment (aabadie) [inline sys/include/net/sock/ip.h]: utilization => usage, in form of => like

File: sys/include/net/sock/ip.h [hunk 8]
```diff
@@ -12,7 +12,244 @@
...
+ *     return 0;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Again: Don't forget to also @ref including-modules "include" the IPv6 module
+ * of your networking implementation (e.g. `gnrc_ipv6_default` for
+ * @ref net_gnrc GNRC) and at least one network device.
+ *
+ * We first create again a `sock` with a local end point bound to any IPv6
+ * address. Note that we also could set the remote here and not use it with
```
Comment (aabadie) [inline sys/include/net/sock/ip.h]: maybe specify remote end point ?

File: sys/include/net/sock/ip.h [hunk 9]
```diff
@@ -12,7 +12,244 @@
...
+ *     sock_ip_ep_t local = SOCK_IPV6_EP_ANY;
+ *     sock_ip_t sock;
+ *
+ *     if (sock_ip_create(&sock, &local, NULL, PROTNUM_IPV6_NONXT, 0) < 0) {
+ *         puts("Error creating raw IP sock");
+ *         return 1;
+ *     }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * We then create a remote end point for the link-local all nodes multicast
```
Comment (aabadie) [inline sys/include/net/sock/ip.h]: not sure I correctly understand this.

File: sys/include/net/sock/ip.h [hunk 10]
```diff
@@ -69,21 +306,21 @@ typedef struct sock_ip sock_ip_t;
...
  * @return  -EAFNOSUPPORT, if `local != NULL` or `remote != NULL` and
  *          sock_ip_ep_t::family of @p local or @p remote is not supported.
  * @return  -EINVAL, if `proto` is not supported or if sock_ip_ep_t::netif of
  *          @p local or @p remote is not a valid interface or contradict each
  *          other (i.e. `(local->netif != remote->netif) &&
  *          ((local->netif != SOCK_ADDR_ANY_NETIF) ||
  *          (remote->netif != SOCK_ADDR_ANY_NETIF))` if neither is `NULL`).
- * @return  -ENOMEM, if the stack can't provide enough resources for `sock` to
- *          be created.
+ * @return  -ENOMEM, if not enough resources for `sock` to be created can be
```
Comment (aabadie) [inline sys/include/net/sock/ip.h]: strange rephrasing ;)

Comment (aabadie) [inline sys/include/net/sock/ip.h]: maybe ... not enough resources can be provided for `sock` to be created ?

File: sys/include/net/sock/udp.h [hunk 11]
```diff
@@ -13,6 +13,245 @@
...
+ * `buf` to store the data received by the server:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * #include "net/af.h"
+ * #include "net/sock/ip.h"
+ *
+ * uint8_t buf[128];
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * To be able to listen for incoming packets we want to bind the sock so we need
```
Comment (aabadie) [inline sys/include/net/sock/udp.h]: we want to bind the sock so we need, etc
=> we bind the sock by setting a local end point with a port ?

File: sys/include/net/sock/udp.h [hunk 12]
```diff
@@ -13,6 +13,245 @@
...
+ *
+ *     local.port = 12345;
+ *
+ *     if (sock_udp_create(&sock, &local, NULL, 0) < 0) {
+ *         puts("Error creating UDP sock");
+ *         return 1;
+ *     }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * We then wait indefinitely for an incoming message in `buf` from `remote`.
```
Comment (aabadie) [inline sys/include/net/sock/udp.h]: I think using 'we' when explaining the behavior of the code should be replaced with something more non personal.
Example:
The program then wait indefinitely for an incoming message...

File: sys/include/net/sock/udp.h [hunk 13]
```diff
@@ -13,6 +13,245 @@
...
+ * one line:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * res = sock_udp_send(NULL, data, data_len, &remote);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * With `data` being the data send, `data_len` the length of `data` and `remote`
+ * the remote end point the packet is to be sent.
+ *
+ * To see some other utilization we look at a more complex example in form of
```
Comment (aabadie) [inline sys/include/net/sock/udp.h]: usage

File: sys/include/net/sock/ip.h (hunk 7 above)
Comment (miri64) [inline sys/include/net/sock/ip.h]: I disagree with usage here. I'm showing off the capabilities here, not its usage (which was already shown in the one-liner above)

File: sys/include/net/sock/ip.h (hunk 9 above)
Comment (miri64) [inline sys/include/net/sock/ip.h]: If you tell me how you understand it I maybe can clear that up. As is the sentence seems straight forward to me... :-/

File: sys/include/net/sock/ip.h (hunk 7 above)
Comment (aabadie) [inline sys/include/net/sock/ip.h]: indeed, in french there's the same word (utilisation), which is like usage. Maybe capacities or capabilities would be better ?

File: sys/include/net/sock/ip.h (hunk 9 above)
Comment (aabadie) [inline sys/include/net/sock/ip.h]: if I understand well (I hope I do ;) ), I would rephrase like this : 
create a remote end point with the link-local multicast address of all nodes and send

File: sys/include/net/sock/udp.h (hunk 13 above)
Comment (miri64) [inline sys/include/net/sock/udp.h]: See https://github.com/RIOT-OS/RIOT/pull/5884#discussion_r80923591

**Comment by miri64** [comment]
Applied requested changes.

File: sys/include/net/sock/ip.h (hunk 9 above)
Comment (miri64) [inline sys/include/net/sock/ip.h]: Better now?

File: sys/include/net/sock.h [hunk 14]
```diff
@@ -34,39 +33,55 @@
...
  *
- * Each network stack must implement at least one sock type.
- *
- * Note that there might be no relation between the different sock types.
- * For simplicity and modularity this API doesn't put any restriction of the
- * actual implementation of the type. For example, one implementation might
- * choose to have all sock types have a common base class or use the raw IP
- * sock type to send e.g. UDP packets, while others will keep them
- * completely separate from each other.
+ * Note that there might be no relation between the different `sock` types.
```
Comment (kaspar030) [inline sys/include/net/sock.h]: Do you think users might try that?

File: sys/include/net/sock.h [hunk 15]
```diff
@@ -34,39 +33,55 @@
...
  *
  * This allows for network stack agnostic code on the application layer.
  * The application code to establish a connection is always the same, allowing
  * the network stack underneath to be switched simply by changing the
  * `USEMODULE` definitions in the application's Makefile.
  *
+ * The actual code very much depends on the used `sock` type. Please refer to
+ * their documentation for specific examples.
+ *
+ * Implementors Notes
```
Comment (kaspar030) [inline sys/include/net/sock.h]: Implementors -> implementor?

File: sys/include/net/sock/ip.h (hunk 9 above)
Comment (aabadie) [inline sys/include/net/sock/ip.h]: yes, thanks for the clarification :)

File: sys/include/net/sock.h (hunk 14 above)
Comment (miri64) [inline sys/include/net/sock.h]: People tried it with `conn` so why wouldn't they try it with `sock`?

File: sys/include/net/sock.h (hunk 15 above)
Comment (miri64) [inline sys/include/net/sock.h]: This is a caption...  Why should I remove the capitalization?

**Comment by kaspar030** [comment]
>> Implementors Notes
> 
> This is a caption... Why should I remove the capitalization?

just remove the plural s at the end of Implementors, capital I is fine.

**Comment by kaspar030** [comment]
> People tried it with conn so why wouldn't they try it with sock?

Someone tried to re-cast the objects into each other?

**Comment by miri64** [comment]
> just remove the plural s at the end of Implementors, capital I is fine.

oops I actually overlooked that.

**Comment by miri64** [comment]
>> People tried it with conn so why wouldn't they try it with sock?
>
> Someone tried to re-cast the objects into each other?

Yes some testers get funny ideas when you do not explicitly state that a behavior is forbidden ;-). And actually I can understand that when you come from the socket world. Also: casting `sock_udp_t` to `gnrc_ip_t` and vice-versa makes also sense to a degree.

**Comment by miri64** [comment]
Mind you, they didn't cast a `conn` that was already established, but used the same type for different operations, after the previous operation was finished.

File: sys/include/net/sock.h (hunk 15 above)
Comment (miri64) [inline sys/include/net/sock.h]: Fixed.

**Comment by immesys** [comment]
I just tried testing this and I get a missing "sock_types.h' from udp.h:198, I have gnrc_sock_udp and gnrc_ipv6_default included. I also cannot find sock_types.h anywhere in the codebase. Any tips?

**Comment by immesys** [comment]
My humblest apologies, I did not see #5772.

**Comment by miri64** [comment]
Ping?

**Comment by miri64** [comment]
@aabadie was that an ACK?

**Comment by aabadie** [comment]
@miri64, Murdock is asking for squashing. Then I'll give an ACK.

**Comment by miri64** [comment]
Squashed.

**Comment by miri64** [comment]
Murdock is happy :grinning:

**Comment by aabadie** [comment]
ACK and Go !
